onnxscript/function_libs/torch_aten/graph_building.py

```diff
@@ -0,0 +1,446 @@
+__all__ = [
+    "TorchScriptTensor",
+    "TorchScriptGraph",
+    "TorchScriptEvaluator",
+]
+
+
+ValidArgumentType = Union["TorchScriptTensor", str, int, float, bool]
+
+# TODO(justinchuby): Build a context manager to handle source information.
+
+
+class TorchScriptTensor(onnxscript_tensor.Tensor):
+    """A onnxscript tensor that wraps a torchscript Value."""
+
+    # TODO(justinchuby): Create setters for attributes so the exporter can
+    # update the name/shape/type information when they are available.
+
+    def __init__(self, value: torch.Value):
+        super().__init__(None)
+        self._value = value
+
+    @property
+    def value(self) -> np.ndarray:
+        return None
+
+    @property
+    def rank(self) -> int | None:
+        value_type = self._value.type()
+        if value_type is None:
+            return None
+        value_type = typing.cast(torch.TensorType, value_type)
+        return value_type.dim()
+
+    @property
+    def shape(self) -> tuple[int | None, ...] | None:
+        value_type = self._value.type()
+        if value_type is None:
+            return None
+        value_type = typing.cast(torch.TensorType, value_type)
+        shape = value_type.varyingSizes()
+        if shape is None:
+            return None
+        return tuple(shape)
+
+    @property
+    def dtype(self):
+        # TODO: Return numpy dtype
+        return _type_utils.JitScalarType.from_value(
+            self._value, default=_type_utils.JitScalarType.UNDEFINED
+        ).dtype()
+
+    @property
+    def onnx_dtype(self):
+        return _type_utils.JitScalarType.from_value(
+            self._value, _type_utils.JitScalarType.UNDEFINED
+        ).onnx_type()
+
+    def symbolic_value(self) -> torch.Value:
+        """The symbolic Value in torch.Graph."""
+        return self._value
+
+
+@beartype
+def _split_args_kwargs_to_input_attr(
+    onnx_func: onnxscript.OnnxFunction, args, kwargs
+) -> tuple[list[ValidArgumentType], dict[str, ValidArgumentType]]:
+    """Split the function args and kwargs to onnx inputs and attributes.
+
+    Args:
+        onnx_func: The onnx function.
+        args: The positional arguments.
+        kwargs: The keyword arguments.
+
+    Returns:
+        A tuple of (onnx_inputs, onnx_attributes).
+    """
+    function_ir = onnx_func.function_ir
+    # FIXME(justinchuby): There can be attributes in the args too.
+    # The first len(func_ir.inputs) arguments are onnx inputs
+    onnx_inputs = args[: len(function_ir.inputs)]
+    # The rest is onnx attributes
+    attributes_in_args = args[len(function_ir.inputs) :]
+    # Construct a dictionary of attributes with their names specified in the function
+    # definition
+    onnx_attributes = {}
+
+    # (1) Some/All attributes are supplied as positional arguments
+    attr_name_to_protos = collections.OrderedDict(
+        (attr.name, attr) for attr in function_ir.attr_protos
+    )
+
+    assert len(function_ir.attr_protos) >= len(attributes_in_args)
```

Comment:
Don't understand what this is checking. In function, attributes are split into those with and without default values. In the args, they are split into args and kwargs. So, what is this checking?

Reply:
Similar questions about the logic below (unless function_ir was changed recently in a way I am unaware of).

Reply:
so some attributes are in args and some are in kwargs. `len(function_ir.attr_protos)` is the number of total attributes. We are asserting that the number of total attributes is more than or equal to the number of attributes in args.

Reply:
AFAIK, `len(function_ir.attrs) + len(function_ir.attr_protos)` is the number of total attributes

Reply:
hmm. how do we get the type for attrs? I thought attr_protos contains everything